app/src/main/java/com/asfoundation/wallet/EwtAuthenticatorService.kt

```diff
@@ -0,0 +1,64 @@
+package com.asfoundation.wallet
+
+import com.appcoins.wallet.bdsbilling.WalletService
+import com.asfoundation.wallet.util.convertToBase64
+import com.google.gson.JsonObject
+
+
+private const val TTL = 3600
+
+class EwtAuthenticatorService(private val walletService: WalletService,
+                              private val header: String) {
+
+  private var cachedAuth: MutableMap<String, Pair<String, Long>> = HashMap()
+
+  @Synchronized
+  fun getEwtAuthentication(address: String, currentUnixTime: Long): String {
+    return if (shouldBuildEwtAuth(address, currentUnixTime))
+      getNewEwtAuthentication(address, currentUnixTime)
+    else {
+      cachedAuth[address]!!.first
+    }
+  }
+
+  @Synchronized
+  fun getNewEwtAuthentication(address: String, currentUnixTime: Long): String {
+    val payload = replaceInvalidCharacters(getPayload(address, currentUnixTime))
+    val signedPayload = walletService.signContent(payload)
+        .blockingGet()
+    return buildAndSaveEwtString(replaceInvalidCharacters(header.convertToBase64()), payload,
+        signedPayload, address,
+        currentUnixTime)
+  }
+
+  private fun shouldBuildEwtAuth(address: String, currentUnixTime: Long): Boolean {
```

Comment:
should the current timestamp be passed as argument or checked to avoid being tricked by the entity calling the method?

Reply:
**Discuss**